### PR TITLE
Initial commit providing openstack functionality for rosco

### PR DESCRIPTION
**Pull request by drmaas**

### CODE REVIEW

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 1]
```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2015 Google, Inc.
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: Copyright is off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,679 @@
+/*
+ * Copyright 2015 Google, Inc.
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Copyright is off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+  String configDir = "/some/path"
+
+  @Shared
+  RoscoOpenstackConfiguration.OpenstackBakeryDefaults openstackBakeryDefaults
+
+  @Shared
+  RoscoConfiguration roscoConfiguration
+
+  void setupSpec() {
+    def openstackBakeryDefaultsJson = [
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Please make the indentation here and everywhere match the neighbors: https://github.com/spinnaker/rosco/blob/master/rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy#L55

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy [hunk 4]
```diff
@@ -56,14 +56,15 @@ class BakeRequest {
   String ami_name
   String ami_suffix
   Boolean upgrade
+  String flavor //openstack only
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy]: Would like to try to avoid introducing (more) provider-specific attributes in common areas. Please change this to something generic like `machine_type` or 'instance_type`, and drop the comment. This seems to be the intent of this attribute based on the description here: http://docs.openstack.org/openstack-ops/content/flavors.html

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/retrofit/RetrofitConfiguration.groovy [hunk 5]
```diff
@@ -36,7 +36,7 @@ class RetrofitConfiguration {
   }
 
   @Bean LogLevel retrofitLogLevel() {
-    LogLevel.FULL
+    LogLevel.BASIC
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/retrofit/RetrofitConfiguration.groovy]: Please leave the log level as it was originally. Will be dropping this fairly shortly anyway, and will reduce the number of uncommitted changes in the meantime.

File: rosco-web/config/rosco.yml [hunk 6]
```diff
@@ -4,7 +4,7 @@ server:
 rush:
   baseUrl: http://localhost:8085
   # TODO(duftler): Figure out the right way to set this dynamically.
-  configDir: /some/path/to/config/packer
+  configDir: rosco-web/config/packer
```
Comment (duftler) [inline rosco-web/config/rosco.yml]: Did this change make things work for you? Pretty sure that this config value is used in building the command to rush. And then rush will try to resolve that path, not rosco.

File: rosco-web/config/rosco.yml [hunk 7]
```diff
@@ -191,6 +191,34 @@ google:
         virtualizationSettings:
           sourceImage: ubuntu-1404-trusty-v20150909a
 
+openstack:
+  enabled: ${OS_ENABLED:false}
+  bakeryDefaults:
+    identityEndpoint: ${OS_AUTH_URL:https://openstack.yourcompany.com:5000/v3}
+    domainName: Default
+    floatingIpPool: ext_vlan1472_net
```
Comment (duftler) [inline rosco-web/config/rosco.yml]: Is this a special value?

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy [hunk 8]
```diff
@@ -78,12 +78,18 @@ class BakeryController {
...
   }
 
-  @ExceptionHandler
+  @ExceptionHandler(BakeOptions.Exception)
   @ResponseStatus(HttpStatus.NOT_FOUND)
   Map handleBakeOptionsException(BakeOptions.Exception e) {
     [error: "bake.options.not.found", status: HttpStatus.NOT_FOUND, messages: ["Bake options not found. " + e.message]]
   }
 
+  @ExceptionHandler(Exception)
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: What's the reason for this change?

I don't know enough spring boot to understand the implications without more context.

How does this change what is returned in the case of an exception? Do any changes need to be made to the callers of this entry-point?

File: rosco-web/config/rosco.yml (hunk 7 above)
Comment (drmaas) [inline rosco-web/config/rosco.yml]: no, will remove the env ref

File: rosco-web/config/rosco.yml (hunk 6 above)
Comment (drmaas) [inline rosco-web/config/rosco.yml]: did not test yet - will revert

File: rosco-web/config/rosco.yml [hunk 9]
```diff
@@ -191,6 +191,34 @@ google:
...
+    securityGroups: default
+    tenantName: PRD-spinnakeropenstack
```
Comment (duftler) [inline rosco-web/config/rosco.yml]: Maybe this one too?

Comment (drmaas) [inline rosco-web/config/rosco.yml]: If you're referring to username/password, I'm not sure how we want to manage it. This way allows it to work in a reasonably secure way ... but there might be a better way.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy (hunk 8 above)
Comment (drmaas) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: It was for debugging.

 I actually think we should have better exception handling. I found that I got generic bad request messages with null message text until I added the above.

File: rosco-web/config/rosco.yml (hunk 9 above)
Comment (duftler) [inline rosco-web/config/rosco.yml]: Sorry, I was just asking if `PRD-spinnakeropenstack` was a special value that you intended to commit. Or a placeholder, or something else.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy (hunk 8 above)
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: Please file an issue on https://github.com/spinnaker/spinnaker/issues with some steps to reproduce the null message text and we'll fix it as a separate pr. Will need to test any fix with the full spinnaker stack, so it'll be easier to tackle it independently.

File: rosco-web/config/rosco.yml (hunk 9 above)
Comment (drmaas) [inline rosco-web/config/rosco.yml]: no, it's just a dummy value.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy (hunk 8 above)
Comment (drmaas) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: https://github.com/spinnaker/spinnaker/issues/894

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+  @Override
+  def getBakeryDefaults() {
+    return openstackBakeryDefaults
+  }
+
+  @Override
+  BakeOptions getBakeOptions() {
+    new BakeOptions(
+            cloudProvider: BakeRequest.CloudProviderType.openstack,
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: Indentation looks off here.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+      openstackVirtualizationSettings.sourceImageName = bakeRequest.base_ami
+    }
+
+    return openstackVirtualizationSettings
+  }
+
+  @Override
+  Map buildParameterMap(String region, def openstackVirtualizationSettings, String imageName, BakeRequest bakeRequest) {
+    def parameterMap = [
+            openstack_identity_endpoint   : openstackBakeryDefaults.identityEndpoint,
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: Indentation looks off here.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+            openstack_region              : region,
+            openstack_ssh_username        : openstackVirtualizationSettings.sshUserName,
+            openstack_instance_type       : openstackVirtualizationSettings.instanceType,
+            openstack_source_image_name   : openstackVirtualizationSettings.sourceImageName,
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: Don't need the extra spaces before the colon.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy [hunk 13]
```diff
@@ -465,7 +465,7 @@ class GCEBakeHandlerSpec extends Specification {
       def packerCommandFactoryMock = Mock(PackerCommandFactory)
       def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
                                         package_name: PACKAGES_NAME,
-                                        base_os: "wily",
+                                      base_os: "wily",
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy]: Indentation looks off here.

Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy]: This looks accidental.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,679 @@
...
+  }
+
+  void 'can identify openstack builds'() {
+    setup:
+    @Subject
+    OpenstackBakeHandler openstackBakeHandler = new OpenstackBakeHandler(openstackBakeryDefaults: openstackBakeryDefaults)
+
+    when:
+    def logsContent =
+            "    openstack: Discovering enabled extensions...\n"
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,679 @@
...
+  }
+
+  void 'rejects non openstack builds'() {
+    setup:
+    @Subject
+    OpenstackBakeHandler openstackBakeHandler = new OpenstackBakeHandler(openstackBakeryDefaults: openstackBakeryDefaults)
+
+    when:
+    def logsContent =
+            "    somesystem-thing: doing the harlem shake ...\n"
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,679 @@
...
+    }
+  }
+
+  void 'produces packer command with all required parameters'() {
+    setup:
+    String instanceType = openstackBakeryDefaults.baseImages[0].virtualizationSettings[0].instanceType
+    def imageNameFactoryMock = Mock(ImageNameFactory)
+    def packerCommandFactoryMock = Mock(PackerCommandFactory)
+    def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
+            package_name: PACKAGES_NAME,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 17]
```diff
@@ -0,0 +1,679 @@
...
+    1 * packerCommandFactoryMock.buildPackerCommand("", parameterMap, "$configDir/$openstackBakeryDefaults.templateFile")
+  }
+
+  void 'produces packer command with all required parameters, overriding base ami'() {
+    setup:
+    String instanceType = openstackBakeryDefaults.baseImages[0].virtualizationSettings[0].instanceType
+    def imageNameFactoryMock = Mock(ImageNameFactory)
+    def packerCommandFactoryMock = Mock(PackerCommandFactory)
+    def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
+            package_name: PACKAGES_NAME,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 18]
```diff
@@ -0,0 +1,679 @@
...
+  }
+
+  void 'produces packer command with all required parameters, overriding template filename'() {
+    setup:
+    String template_file_name = "somePackerTemplate.json"
+    String instanceType = openstackBakeryDefaults.baseImages[0].virtualizationSettings[0].instanceType
+    def imageNameFactoryMock = Mock(ImageNameFactory)
+    def packerCommandFactoryMock = Mock(PackerCommandFactory)
+    def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
+            package_name: PACKAGES_NAME,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 19]
```diff
@@ -0,0 +1,679 @@
...
+            openstack_tenant_name: openstackBakeryDefaults.tenantName,
+            repository: DEBIAN_REPOSITORY,
+            package_type: BakeRequest.PackageType.DEB.packageType,
+            packages: PACKAGES_NAME,
+            configDir: configDir
+    ]
+
+    @Subject
+    OpenstackBakeHandler openstackBakeHandler = new OpenstackBakeHandler(configDir: configDir,
+            openstackBakeryDefaults: openstackBakeryDefaults,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 20]
```diff
@@ -0,0 +1,679 @@
...
+    1 * packerCommandFactoryMock.buildPackerCommand("", parameterMap, "$configDir/$template_file_name")
+  }
+
+  void 'produces packer command with all required parameters, adding extended attributes'() {
+    setup:
+    String instanceType = openstackBakeryDefaults.baseImages[0].virtualizationSettings[0].instanceType
+    def imageNameFactoryMock = Mock(ImageNameFactory)
+    def packerCommandFactoryMock = Mock(PackerCommandFactory)
+    def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
+            package_name: PACKAGES_NAME,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 21]
```diff
@@ -0,0 +1,679 @@
...
+            package_type: BakeRequest.PackageType.DEB.packageType,
+            packages: PACKAGES_NAME,
+            configDir: configDir,
+            someAttr1: "someValue1",
+            someAttr2: "someValue2"
+    ]
+
+    @Subject
+    OpenstackBakeHandler openstackBakeHandler = new OpenstackBakeHandler(configDir: configDir,
+            openstackBakeryDefaults: openstackBakeryDefaults,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy [hunk 22]
```diff
@@ -0,0 +1,679 @@
...
+    1 * packerCommandFactoryMock.buildPackerCommand("", parameterMap, "$configDir/$openstackBakeryDefaults.templateFile")
+  }
+
+  void 'produces packer command with all required parameters, and overrides native attributes via extended attributes'() {
+    setup:
+    String instanceType = openstackBakeryDefaults.baseImages[0].virtualizationSettings[0].instanceType
+    def imageNameFactoryMock = Mock(ImageNameFactory)
+    def packerCommandFactoryMock = Mock(PackerCommandFactory)
+    def bakeRequest = new BakeRequest(user: "someuser@gmail.com",
+            package_name: PACKAGES_NAME,
```
Comment (duftler) [inline rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy]: Indentation looks off here.

File: rosco-web/config/packer/openstack.json [hunk 23]
```diff
@@ -0,0 +1,51 @@
+{
+  "variables": {
+    "openstack_domain_name": null,
+    "openstack_instance_type": null,
+    "openstack_floating_ip_pool": null,
+    "openstack_image_name": null,
+    "openstack_identity_endpoint": null,
+    "openstack_insecure": "false",
+    "openstack_source_image_name": null,
+    "openstack_ssh_username":null,
```
Comment (duftler) [inline rosco-web/config/packer/openstack.json]: Missing a space before the `null`.

File: rosco-web/config/packer/openstack.json [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+    "openstack_username": "{{env `OS_USERNAME`}}",
+    "openstack_password": "{{env `OS_PASSWORD`}}",
+    "openstack_region": null,
+    "openstack_security_groups":null,
```
Comment (duftler) [inline rosco-web/config/packer/openstack.json]: Missing a space before the `null`.

File: rosco-web/config/packer/openstack.json [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    "openstack_tenant_name":null,
```
Comment (duftler) [inline rosco-web/config/packer/openstack.json]: Missing a space before the `null`.

File: rosco-web/config/packer/openstack.json [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+  "builders": [{
+    "type": "openstack",
+    "domain_name": "{{user `openstack_domain_name`}}",
+    "flavor": "{{user `openstack_instance_type`}}",
+    "floating_ip_pool": "{{user `openstack_floating_ip_pool`}}",
+    "image_name": "{{user `openstack_image_name`}}",
+    "identity_endpoint": "{{user `openstack_identity_endpoint`}}",
+    "insecure": "{{user `openstack_insecure`}}",
+    "region": "{{user `openstack_region`}}",
+    "ssh_username":"{{user `openstack_ssh_username`}}",
```
Comment (duftler) [inline rosco-web/config/packer/openstack.json]: Missing a space following the colon.

**Comment by drmaas** [comment]
I suggest providing a groovy code formatting template for people to follow. I was able to figure out the intellij options to format according to the existing conventions, but it took a bit.

I'm going to add it to the README PR.

**Comment by duftler** [comment]
The main goal is to just be consistent with the neighbors. The primary authors of each component have their own styles, and new contributors are expected to try to be as consistent as possible.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy [hunk 27]
```diff
@@ -78,7 +85,7 @@ class BakeryController {
     return baseImage
   }
 
-  @ExceptionHandler
+  @ExceptionHandler(BakeOptions.Exception)
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: What is the effect of this change?

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy [hunk 28]
```diff
@@ -32,7 +32,14 @@ import io.swagger.annotations.ApiParam
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.beans.factory.annotation.Value
 import org.springframework.http.HttpStatus
-import org.springframework.web.bind.annotation.*
+import org.springframework.web.bind.annotation.ExceptionHandler
```
Comment (duftler) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: What's the reason to replace the wildcard import with explicit imports?

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy [hunk 29]
```diff
@@ -56,14 +56,15 @@ class BakeRequest {
   String ami_name
   String ami_suffix
   Boolean upgrade
+  String instance_type
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy]: I notice that the packer template treats this new attribute as required. It will take a bit of extra work in orca (the orchestration engine) to allow you to pass this new attribute from a pipeline bake stage all the way through to rosco. The reason for the extra work is that the internal Netflix bakery (which is not Rosco, but is largely interface-compatible with Rosco) will not accept requests with extraneous attributes. So the new attribute will need to be explicitly dropped from requests that are not routed to Rosco. I'll grab a link so you have it for later when you get to adding ui/orchestration support.

Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/api/BakeRequest.groovy]: So at some point, you'll want to add this same `instance_type` attribute to orca's `BakeRequest` here: https://github.com/spinnaker/orca/blob/master/orca-bakery/src/main/groovy/com/netflix/spinnaker/orca/bakery/api/BakeRequest.groovy#L30

That will allow orca to unmarshall it in requests coming from the ui (deck). Then, you'll need to ensure that it only gets sent to Rosco, and is not included on bake requests propagated from orca to the Netflix internal bakery: https://github.com/spinnaker/orca/blob/master/orca-bakery/src/main/groovy/com/netflix/spinnaker/orca/bakery/tasks/CreateBakeTask.groovy#L146

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+import com.netflix.spinnaker.rosco.api.BakeOptions
+import com.netflix.spinnaker.rosco.api.BakeRequest
+import com.netflix.spinnaker.rosco.providers.CloudProviderBakeHandler
+import com.netflix.spinnaker.rosco.providers.openstack.config.RoscoOpenstackConfiguration
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+
+/**
+ * TODO need to figure out:
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: Please add a name to the TODO: `TODO(drmaas)`

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  @Override
+  String produceProviderSpecificBakeKeyComponent(String region, BakeRequest bakeRequest) {
+    region
+  }
+
+  @Override
+  def findVirtualizationSettings(String region, BakeRequest bakeRequest) {
+
```
Comment (duftler) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandler.groovy]: There is an extraneous blank line here.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy (hunk 27 above)
Comment (drmaas) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: None - it is for clarity.

File: rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy (hunk 28 above)
Comment (drmaas) [inline rosco-web/src/main/groovy/com/netflix/spinnaker/rosco/controllers/BakeryController.groovy]: None - will undo.

File: build.gradle [hunk 32]
```diff
@@ -45,7 +45,9 @@ allprojects {
       exceptionFormat = 'full'
     }
   }
+  tasks.withType(JavaExec) {
```
Comment (duftler) [inline build.gradle]: What is this change needed for? I may have missed it, but I didn't see anything referencing the propagated properties.

**Comment by duftler** [comment]
LGTM after addressing one remaining comment (re: the propagation of the system properties).

**Comment by duftler** [comment]
This PR will need a rebase before I can merge.

File: build.gradle (hunk 32 above)
Comment (drmaas) [inline build.gradle]: It was for me to be able to do something like this during development:
```
./gradlew -Drush.configDir=/Users/z001mj7/Projects/spinnaker/rosco/rosco-web/config/packer bootRun
```

Comment (duftler) [inline build.gradle]: You should be able to accomplish the same thing by setting an env variable `RUSH_CONFIG_DIR`. There are some more details here: http://www.spinnaker.io/docs/custom-configuration#section-about-notation-used-within-the-yaml-files

I think once this block is removed, and the PR rebased, we're ready to merge.

**Comment by duftler** [comment]
LGTM.
